Migration/tests

ronyeh opened the pull request:
This PR improves the tests while revealing ways we can improve (or fix) the TypeScript typings in VexFlow.

Jump down to the most recent discussion: https://github.com/0xfe/vexflow/pull/1094#issuecomment-906897447

---

Here is the entire list of 67 modified files:
```
tests/accidental_tests.ts
tests/annotation_tests.ts
tests/articulation_tests.ts
tests/auto_beam_formatting_tests.ts
tests/bach_tests.ts
tests/barline_tests.ts
tests/beam_tests.ts
tests/bend_tests.ts
tests/boundingbox_tests.ts
tests/boundingboxcomputation_tests.ts
tests/chordsymbol_tests.ts
tests/clef_tests.ts
tests/curve_tests.ts
tests/dot_tests.ts
tests/easyscore_tests.ts
tests/factory_tests.ts
tests/formatter_tests.ts
tests/fraction_tests.ts
tests/ghostnote_tests.ts
tests/glyphnote_tests.ts
tests/gracenote_tests.ts
tests/gracetabnote_tests.ts
tests/key_clef_tests.ts
tests/keymanager_tests.ts
tests/keysignature_tests.ts
tests/mocks.ts
tests/modifier_tests.ts
tests/multimeasurerest_tests.ts
tests/music_tests.ts
tests/notehead_tests.ts
tests/notesubgroup_tests.ts
tests/ornament_tests.ts
tests/parser_tests.ts
tests/pedalmarking_tests.ts
tests/percussion_tests.ts
tests/registry_tests.ts
tests/rests_tests.ts
tests/rhythm_tests.ts
tests/stave_tests.ts
tests/staveconnector_tests.ts
tests/stavehairpin_tests.ts
tests/staveline_tests.ts
tests/stavemodifier_tests.ts
tests/stavenote_tests.ts
tests/stavetie_tests.ts
tests/stringnumber_tests.ts
tests/strokes_tests.ts
tests/style_tests.ts
tests/tabnote_tests.ts
tests/tabslide_tests.ts
tests/tabstave_tests.ts
tests/tabtie_tests.ts
tests/textbracket_tests.ts
tests/textnote_tests.ts
tests/threevoice_tests.ts
tests/tickcontext_tests.ts
tests/timesignature_tests.ts
tests/tuning_tests.ts
tests/tuplet_tests.ts
tests/typeguard_tests.ts
tests/types/qunit.d.ts
tests/vexflow_test_helpers.ts
tests/vf_prefix_tests.ts
tests/vibrato_tests.ts
tests/vibratobracket_tests.ts
tests/voice_tests.ts

tools/generate_png_images.js
```

rvilarl commented:
@ronyeh this PR is getting very big, I suggest that you do other PRs with parts of this one. Once merged you can rebase them into this one reducing the number of files. 
A topic could be note to tickable in voice.

ronyeh commented:
Agreed. I will try to break up this PR into smaller, more digestible ones.

ronyeh commented:
Done! Separated into multiple PRs.

This PR now only touches test related files:

```
tests/accidental_tests.ts
tests/annotation_tests.ts
tests/auto_beam_formatting_tests.ts
tests/bach_tests.ts
tests/barline_tests.ts
tests/beam_tests.ts
tests/bend_tests.ts
tests/boundingboxcomputation_tests.ts
tests/chordsymbol_tests.ts
tests/clef_tests.ts
tests/curve_tests.ts
tests/dot_tests.ts
tests/easyscore_tests.ts
tests/factory_tests.ts
tests/formatter_tests.ts
tests/fraction_tests.ts
tests/ghostnote_tests.ts
tests/glyphnote_tests.ts
tests/gracenote_tests.ts
tests/gracetabnote_tests.ts
tests/key_clef_tests.ts
tests/keysignature_tests.ts
tests/mocks.ts
tests/modifier_tests.ts
tests/multimeasurerest_tests.ts
tests/music_tests.ts
tests/notehead_tests.ts
tests/notesubgroup_tests.ts
tests/ornament_tests.ts
tests/parser_tests.ts
tests/percussion_tests.ts
tests/registry_tests.ts
tests/rests_tests.ts
tests/rhythm_tests.ts
tests/stave_tests.ts
tests/staveconnector_tests.ts
tests/staveline_tests.ts
tests/stavemodifier_tests.ts
tests/stavenote_tests.ts
tests/stavetie_tests.ts
tests/stringnumber_tests.ts
tests/strokes_tests.ts
tests/style_tests.ts
tests/support/qunit_api.ts
tests/tabnote_tests.ts
tests/tabslide_tests.ts
tests/tabstave_tests.ts
tests/tabtie_tests.ts
tests/textnote_tests.ts
tests/threevoice_tests.ts
tests/tickcontext_tests.ts
tests/tuplet_tests.ts
tests/vexflow_test_helpers.ts
tests/vibrato_tests.ts
tests/voice_tests.ts
tools/generate_png_images.js
```

It is rebased off of the other PRs, so we need to review and merge #1112 and then #1116 before doing this one.

ronyeh commented:
No visual diffs from master 12ecf1eee743e15f23b52578f1bc3e5fac7717e2

0xfe commented:
ready for review?

ronyeh commented:
Yes, this one is ready for review. It was based directly off of the
tickables branch.

On Mon, Aug 30, 2021, 2:44 PM Mohit Cheppudira ***@***.***>
wrote:

> ready for review?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/0xfe/vexflow/pull/1094#issuecomment-908722271>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAB2MCLFYQHYC2R3VP2D2H3T7P3VXANCNFSM5BMJHWNQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

ronyeh commented:
I can rebase it off of master so maybe it will be easier to review via the GitHub UI.....

ronyeh commented:
I have rebased this branch off of today's master, but will run the visual diffs again to make sure.

ronyeh commented:
I put together a small wiki page about writing a new tests file.

https://github.com/0xfe/vexflow/wiki/Testing

ronyeh commented:
## Overall Strategy
- Let TypeScript catch API errors for us.
- Minimize/reduce/eliminate code that isn't directly related to the VexFlow API.
- Standardize the structure of a tests file so that it's easier to skim.

## For each test file
- Convert to TypeScript.
- Record API mismatches in a set of TODOs left at the top of each file.
- Add `eslint-disable` and `ts-nocheck` to allow the build process to proceed.



## Specifics
- Each test looks like:
```
const MyTests = {
  Start(): void {
    QUnit.module('Annotation');
    const run = VexFlowTests.runTests;
      run('Test Name', this.myTestMethod);
    ...
  },

  myTestMethod(options: TestOptions): void {
    ...
  },
}

// Helper Functions
....

```
- Move helper functions to the bottom so that all test files start the same way. For bigger blocks of helper code, I have started adding #region markers. This is a [VS Code feature that enables code folding](https://code.visualstudio.com/docs/editor/codebasics#_folding). If you don't like the look of the #region / #endregion markers, I can remove them. :-):
```
//#region Helper Functions
function helper1() {
  ...
}
function helper2() {
  ...
}
function helper3() {
  ...
}
//#endregion
```


- Use arrow functions for callbacks in `.forEach(...)`, `.map(...)`, etc.

ronyeh commented:
OK, `npm run test:reference` shows no diffs relative to today's master at a3ac41ef4041d5d321d10eaf96b1b52857349b58

0xfe commented:
Great, reviewing.

0xfe commented on tests/articulation_tests.ts:
```diff
@@ -256,7 +253,7 @@ const ArticulationTests = {
     Formatter.FormatAndDraw(ctx, staveBar3, notesBar3);
     // bar 4 - juxtaposing second bar next to first bar
-    const staveBar4 = new Stave(staveBar3.width + staveBar3.x, staveBar3.y, 150);
+    const staveBar4 = new Stave(staveBar3.getWidth() + staveBar3.getX(), staveBar3.getY(), 150);
```
Thanks for cleaning up these calls. Curiously did you notice any effect on test performance?

ronyeh replied:
I didn't notice anything regarding performance, especially since these getters are only called a few times.

We would need a larger test, something like a multi-page Bach Demo that uses a lot of getters, to see if it has a noticeable effect on performance.

This discussion does remind me: **Do we always need accessors for simple fields like x, y, width, height?** 

Sometimes it feels excessive, and it might reduce readability in areas of code where we do lots of calculations.

For example, `BoundingBox` has 4 protected members (x, y, w, h). So external code has to use .getX(), .getY(), .getW(), .getH(). BoundingBox is just a entity for doing geometric calculations, so maybe x, y, w, h should all just be public? We can explore this in a separate PR if you want.

0xfe replied:
It's a good point. I typically default to not using accessors for simple fields. But VexFlow is so big now, it's hard to draw the line around what _simple_ means, so we end up with a lot of inconsistency, which raises the cognitive load ("should this be "note.foo" or "note.getFoo()"). So I think it's easier on us to default to using accessors for everything.

0xfe commented on tests/bach_tests.ts:
```diff
@@ -67,7 +73,7 @@ const BachDemoTests = {
 
     /*  Measure 2 */
-    system = makeSystem(150);
+    system = makeSystemAndAdvanceX(150);
```
How about `appendSystem` instead, so we're not so tied to the implementation?

0xfe commented on tests/bend_tests.ts:
```diff
@@ -265,4 +248,11 @@ const BendTests = {
+const note = (tab_struct: TabNoteStruct) => new TabNote(tab_struct);
+const bendWithText = (text: string, release = false) => new Bend(text, release);
+const bendWithPhrase = (phrase: BendPhrase[]) => new Bend('', false, phrase);
+//#endregion Helper Functions
```
What are these `#region/endregion` annotations?

Also, can we move these helpers to the top? It's more readable if dependencies are declared before the they are used.

ronyeh replied:
https://github.com/0xfe/vexflow/pull/1094#issuecomment-908785984

The above comment discusses my overall approach to cleaning up the tests.

#region markers are VS code specific and enable large blocks of code to be folded with a shortcut key (e.g. ctrl+shift+[ on windows / cmd+opt+[ on mac) or by clicking on the left gutter. I have started using them in my recent PRs for large related blocks of code. 

For example in a later PR, I noticed that glyph.ts has 180 lines of static methods before the constructor(). I plan to add #region / #endregion blocks around some of these larger blocks to assist developers with navigating around the code.

https://github.com/0xfe/vexflow/blob/29a838e5d7a940c4b6c74afb2d19741d4ea67fab/src/glyph.ts#L186-190

That said...some of these helper function regions aren't that long. I put them everywhere for consistency's sake, but maybe they should only appear if the region is very long and unwieldy, like over 150-200 lines of code.

Also, they are specific to a particular IDE and are a little ugly (albeit useful!).

If you prefer to not have the #region markers, I can totally remove them.

0xfe replied:
Thanks Ron. I'm not really a fan of the syntax, or the vs-code specific behavior. If we're going to use them, then let's use them sparingly where there's large blocks of clear noise.

ronyeh replied:
Agreed! I also think the syntax looks weird, but it seemed useful while I was extracting some of the helper functions.

I'll remove the #region annotations.

0xfe commented on tests/boundingboxcomputation_tests.ts:
```diff
@@ -235,4 +211,35 @@ const BoundingBoxComputationTests = {
 };
 
+//#region  Helper Functions
+function createCanvas(options: TestOptions): RenderContext {
```
Same comments about moving to top.

ronyeh replied:
> Also, can we move these helpers to the top? It's more readable if dependencies are declared before the they are used.

Yeah, I will move the helper functions up.

I was struggling with making the test files look consistent. For example, I want the `Start()` function to be as close to the top as possible, because I believe it gives a great overview on how many test cases will be run, and what each test case is called.

Some of the test files previously had nested inline callback functions that impaired readability, so I tried to extract those out into helper methods. Since editors and IDEs (e.g., vs code) allow you to quickly follow method invocations (via keyboard shortcut or ctrl+click / cmd+click), I figured it was okay to push the helper functions down.

But yes, I generally agree that hiding helper functions at the bottom is not the best practice. I'll find a way to move them up while keeping the Start() method easy to look for.

0xfe replied:
I see what you're saying about consistency, and I really like the methods you extracted. I think of consistency as a tool for making code more readable. So in this case, the consistency takes away the readability :-)

ronyeh replied:
Got it. I'll move the helpers up. Will ping you when it's ready for a second review.

0xfe commented on tests/gracetabnote_tests.ts:
```diff
@@ -3,39 +3,34 @@
-    const stave = new VF.TabStave(10, 10, x || 350).addTabGlyph().setContext(ctx).draw();
-
-    return { context: ctx, stave: stave };
+    // TODO: Rename tests below since it is redundant with the module name.
```
+1

ronyeh replied:
Re: renaming test cases.

Since I've spent a lot of time in `flow.html` recently, I've noticed that many test cases have redundant naming since we use the Module Name > Test Case Name > Renderer + Font Name.

For example, "Accidental › Accidental Padding › Canvas + Bravura" or "Grace Tab Notes › Grace Tab Note Slurred › SVG + Bravura".

We will need to explore this in a separate PR, since renaming test cases is not trivial.

It **breaks the visual diffing** between versions, since it is done by file name. I haven't thought of a great way to preserve diffing between versions other than having a lookup table to translate test names between versions.

0xfe replied:
Outside of manually adding stable identifiers for each test (or maybe generating them from file + function name), there isn't a better way.

This was something I never anticipated -- we can discuss for a separate PR.

ronyeh commented:
Summary of discussion:
- Move helper functions up so that they are not hidden at the end of test files.
- `#region` markers, yay or nay? They seem unnecessary to use for blocks of code that are < 100 lines, but what about very large blocks of static functions or helper functions? For example, there are over 200 lines of static helper methods between Formatter's class declaration and its constructor. The Glyph class also has a large block of static helper methods before the constructor. They are VS Code specific annotations, but they assist with code folding large blocks of code. Let me know your decision. I'm happy to reserve them just for big blocks of code, or avoid them entirely.

ronyeh commented:
OK, I removed all the #region markers and moved the helper functions up before the test cases that depend on them.

There is a visual diff in one vibrato test, because I adjusted the height from 140px to 240px to be consistent with the other test case, which it is almost identical to:
```
Vibrato.Simple_Vibrato.Bravura 48.5176
Vibrato.Simple_Vibrato.Petaluma 44.9142
Vibrato.Simple_Vibrato.Gonville 28.057
```

BEFORE:
![Vibrato Simple_Vibrato Bravura_Reference](https://user-images.githubusercontent.com/239113/131891253-73f42a01-4e8e-4790-b274-352fa5db307c.png)

AFTER:
![Vibrato Simple_Vibrato Bravura_Current](https://user-images.githubusercontent.com/239113/131891250-67003add-cb39-4114-9f2d-f473e2b8d9cc.png)

I moved the test case functions out to the top level (less indented). Since we are using JS modules, all test functions, helper functions, and helper constants are private by default.

The only thing exported by each test module is the test object with the `Start()`method. Every test file starts with this object. Test files are structured as follows:

```
const VibratoTests = {
  Start(): void {
    QUnit.module('Vibrato');
    const run = VexFlowTests.runTests;
    run('Simple Vibrato', simple);
    run('Harsh Vibrato', harsh);
    run('Vibrato with Bend', withBend);
  },
};

... helper functions and constants ...

function simple(options: TestOptions, contextBuilder: ContextBuilder): void {
  ...
}

function harsh(options: TestOptions, contextBuilder: ContextBuilder): void {
  ...
}

function withBend(options: TestOptions, contextBuilder: ContextBuilder): void {
  ...
}

export { VibratoTests };
```

0xfe commented:
Ron, this is really amazing work and very thoughtfully done! Bravo! I really appreciate your attention to detail here, and the focus on readability and code health. (There's so much technical debt in this code base, and you've done a lot to move us in the right direction!)

Merging.

@mscuthbert @gristow FYI.
